### DISCUSSION TITLE
How to create `TypeVar` that is either a subclass of `A` or a subclass of `B` ?

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1260] [Q&A]
Consider the case when we want a `TypeVar` named `X` that is **either** a subclass of `A`, **or** a subclass of `B`.

> How can this be realized using typing & checking with `mypy` ????

Neither `X=TypeVar("X", bound=Union[A,B])` nor  `X = TypeVar("X", A, B, covariant=True)` seem to do the right thing (see example below).

<details><summary> Minimal Working Example </summary>

```python
from __future__ import annotations
from typing import *

SelfA = TypeVar("SelfA", bound="A")
SelfB = TypeVar("SelfB", bound="B")

# fmt: off
class A:
    def __add__(self: SelfA, other: SelfA) -> SelfA:  return self
    def __radd__(self: SelfA, other: SelfA) -> SelfA: return self

class B:
    def __add__(self: SelfB, other: SelfB) -> SelfB:  return self
    def __radd__(self: SelfB, other: SelfB) -> SelfB: return self

class SubA(A): ...
class SubB(B): ...


T = TypeVar("T", bound=A)
class Foo(Generic[T]):
    s: T
    def __init__(self, x: T, y: T):
        self.s = x + y                   # ✔

foo_a: Foo[A] = Foo(A(), A())            # ✔
foo_b: Foo[A] = Foo(SubA(), A())         # ✔
foo_e: Foo[SubA] = Foo(SubA(), SubA())   # ✔


AuB = TypeVar("AuB", bound=Union[A,B])
class Bar(Generic[AuB]):
    s: AuB
    def __init__(self, x: AuB, y: AuB):
        self.s = x + y                   # ✘ raises type[assignment, operator]

bar_a: Bar[A] = Bar(A(), A())            # ✔
bar_b: Bar[A] = Bar(SubA(), A())         # ✔
bar_e: Bar[SubA] = Bar(SubA(), SubA())   # ✔


AB = TypeVar("AB", A, B, covariant=True)  # covariant=True does nothing?
class Baz(Generic[AB]):
    s: AB
    def __init__(self, x: AB, y: AB):
        self.s = x + y                   # ✔

baz_a: Baz[A] = Baz(A(), A())            # ✔
baz_b: Baz[A] = Baz(SubA(), A())         # ✔
baz_e: Baz[SubA] = Baz(SubA(), SubA())   # ✘ raises type[assignment, type-var]
# fmt: on
```

using `python=3.10.6` and `mypy=0.971`.

</details>

**Comment by erictraut** [comment]
A bound TypeVar won't work in this case if you expect `x + y` to type check because `x` and `y` are both of type `A | B` (or subtypes thereof), and `__add__` is not supported for mismatched combinations of `A` and `B` operands.

A constrained TypeVar should work in this case. Mypy appears to have a bug here on the line that begins with `baz_e:`. This type checks without errors in pyright. (Incidentally, you're using a very old version of mypy. I checked with the latest version, but the bug still persists.)

**Reply by hauntsaninja** [reply]
I may be missing something completely, but mypy doesn't allow value constrained typevars like TypeVar("T", str, bytes) to solve to a subtype of str or bytes. See https://mypy.readthedocs.io/en/stable/generics.html#type-variables-with-value-restriction

I believe this behaviour is specified in PEP 484 which says:

> type constraints cause the inferred type to be _exactly_ one of the constraint types, while an upper bound just requires that the actual type is a subtype of the boundary type. [emphasis from the PEP]

So just saying `Baz[SubA]` is enough to cause mypy to complain with `Value of type variable "AB" of "Baz" cannot be "SubA"`. mypy is happy with e.g. `bar_f: Baz[A] = Baz(SubA(), SubA())`

(also OP is using mypy 0.971, which is latest)

**Reply by erictraut** [reply]
Pyright won't solve constrained TypeVars to a subtype in the absence of any context.

```python
baz_f = Baz(SubA(), SubA())
reveal_type(baz_f) # Baz[A]
```

But in this case there is additional context in the form of the declared type `Baz[SubA]`. If you agree that `Baz[SubA]` is a legal specialization of `Baz`, then it should be possible for a function to return that type, no?

(My bad on the version comment. I thought mypy was on 0.98x already.)

**Reply by hauntsaninja** [reply]
I believe mypy's position is that `Baz[SubA]` is not a legal specialisation. As far as I can tell, mypy is consistent on this point, e.g. it'll complain about just `baz_e: Baz[SubA]` even in the absence of assignment. mypy wants all value constrained typevars to only ever be exactly one of the constraint types.

(As for why this is the case, I'm less sure. I can only say that it seems to very intentionally be the case in mypy and that the authors tried to convey that in PEP 484. Probably comes down to modeling some aspect of str or bytes or operators or Python2 or something)

**Reply by erictraut** [reply]
> it'll complain about just `baz_e: Baz[SubA]` even in the absence of assignment

Ah, I didn't realize that. Yes, then I agree that mypy is being self-consistent here, and it sounds like there was some justification for this. Perhaps @JukkaL or @gvanrossum remember the reasoning. I can't think of any type consistency justification for this limitation, but I could be missing something.

If that limitation is intended, then it appears there's no solution to the problem OP posed.

**Reply by JelleZijlstra** [reply]
I believe mypy is correct. Consider this code:

```python
from typing import AnyStr, Generic

class SubStr(str): ...

class Baz(Generic[AnyStr]):
    s: AnyStr
    def __init__(self, x: AnyStr, y: AnyStr):
        self.s = x + y

baz_e: Baz[SubStr] = Baz(SubStr(), SubStr())
reveal_type(baz_e.s)
```

Pyright throws no errors and infers `SubStr` on the last line, but at runtime, `baz_e.s` is a `str`, not a `SubStr`.

**Reply by erictraut** [reply]
Actually, pyright reveals the type of `baz_e.s` on the last line as `str`, not `SubStr`, so that's consistent with runtime behaviors.

**Reply by randolf-scholz** [reply]
So what to do in this case? Should I open a feature request? Because this is something that actually comes up quite often, at least in the kind of data-wrangling code I'm working with.

**Reply by erictraut** [reply]
Here are some options:
1. Change your code to avoid this usage pattern.
2. Use `# type: ignore` to suppress errors emitted by mypy.
3. File a bug report with mypy and try to convince its maintainers that `Baz[SubStr]` should be supported. Then push for this change to be prioritized or submit a PR yourself.
4. Use pyright to type check your code. (This isn't an option if your code is part of a published library because it then needs to work with all type checkers including mypy.)
5. Propose and champion a change to the Python type system. This could be a clarification of how a constrained TypeVar should work in your code sample above. Or it could involve some new extension to the type system. In either case, the typing-sig email list is probably the best place to start that discussion, since the broader typing community would need to be convinced of its merits.
